### ISSUE TITLE
POST /v1/observations (with obs field values) response is 200 and empty sometimes

### ISSUE THREAD

**Issue by tomsaleeba** [inaturalist/iNaturalistAPI#200]
This is a bug report.

In https://github.com/inaturalist/iNaturalistAPI/issues/199#issuecomment-649848000 you showed me a way to create observations with obs fields included. You said not to use it in prod but professional curiosity means I had to try it out in my dev iNat instance. In doing so, I found some weird behaviour.

At a HTTP status code level, the POST always works with a 200 coming back but sometimes the response is empty (has no body). At first I figured this was just something with how I've set up my iNat instance because I'm making it up as I go along, but I *can* reproduce this in the real/production iNat.

The problem is intermittent but if you take enough samples it seems to show up fairly reliably. Here's how to reproduce. Grab your API JWT and run this bash script, which will attempt to create 6 observations. The exact number doesn't matter but we need to create enough to have a chance of seeing the bug. We create one after another (not parallel), that have 3 obs field values and are added to a project. Oh yeah, I figured out I can supply `project_id` to add to a project in the same request, which is cool.

```bash
jwt=${THE_JWT:?you must pass a JWT}
url=https://api.inaturalist.org/v1/observations

speciesGuess="life"
year=2018

function doCurl {
  lat=$1
  lng=$2
  month=$3
  uuid=1234567e-b9c5-11ea-8d0d${lat}${lng}${month}aaaaa
  cat <<EOF | curl -X POST --include -H Authorization:$jwt -H 'Content-Type: application/json' $url --data @/dev/stdin
{"ignore_photos":true,"project_id":65697,"observation":{"latitude":${lat}.92397888888889,"longitude":${lng}.79083108333333,"observed_on_string":"${year}-${month}-29T04:44:55.310Z","species_guess":"${speciesGuess}","observation_field_values_attributes":{"0":{"observation_field_id":12038,"value":"Terrestrial"},"1":{"observation_field_id":11988,"value":"Exact"},"2":{"observation_field_id":12043,"value":1}},"description":null,"positional_accuracy":null,"captive_flag":false,"geoprivacy":"obscured","uuid":"${uuid}"}}
EOF
}

echo -e '\n\n first'
lat=-36
lng=149
month=01
doCurl $lat $lng $month

echo -e '\n\n second'
lat=-35
lng=148
month=02
doCurl $lat $lng $month

echo -e '\n\n third'
lat=-34
lng=147
month=03
doCurl $lat $lng $month

echo -e '\n\n fourth'
lat=-33
lng=146
month=04
doCurl $lat $lng $month

echo -e '\n\n fifth'
lat=-32
lng=145
month=05
doCurl $lat $lng $month

echo -e '\n\n sixth'
lat=-31
lng=144
month=06
doCurl $lat $lng $month
```

I ran this against the real iNat and here were my results:

```
 first
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 30 Jun 2020 04:36:55 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, Access-Control-Allow-Methods, X-Via
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE
X-Content-Type-Options: nosniff
Vary: Accept-Encoding
Age: 0
X-Cache: MISS
Accept-Ranges: bytes



 second
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 30 Jun 2020 04:36:58 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 39831
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, Access-Control-Allow-Methods, X-Via
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE
X-Content-Type-Options: nosniff
ETag: W/"9b97-JlhBrdpS61ftxBD3mKEpmQtl41s"
Vary: Accept-Encoding
Age: 0
X-Cache: MISS
Accept-Ranges: bytes

{"id":51438795,(snip)...

 third
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 30 Jun 2020 04:37:00 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 36743
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, Access-Control-Allow-Methods, X-Via
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE
X-Content-Type-Options: nosniff
ETag: W/"8f87-JlAUrfitY6Y52enU8keMVs1ANyk"
Vary: Accept-Encoding
Age: 0
X-Cache: MISS
Accept-Ranges: bytes

{"id":51438798,(snip)...

 fourth
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 30 Jun 2020 04:37:03 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, Access-Control-Allow-Methods, X-Via
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE
X-Content-Type-Options: nosniff
Vary: Accept-Encoding
Age: 0
X-Cache: MISS
Accept-Ranges: bytes



 fifth
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 30 Jun 2020 04:37:06 GMT
Content-Type: application/json; charset=utf-8
Transfer-Encoding: chunked
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, Access-Control-Allow-Methods, X-Via
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE
X-Content-Type-Options: nosniff
Vary: Accept-Encoding
Age: 0
X-Cache: MISS
Accept-Ranges: bytes



 sixth
HTTP/1.1 200 OK
Server: nginx
Date: Tue, 30 Jun 2020 04:37:08 GMT
Content-Type: application/json; charset=utf-8
Content-Length: 39817
Connection: keep-alive
Access-Control-Allow-Origin: *
Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization, Access-Control-Allow-Methods, X-Via
Access-Control-Allow-Methods: GET, POST, OPTIONS, PUT, DELETE
X-Content-Type-Options: nosniff
ETag: W/"9b89-0qdBHpxgxPDiH5ZCEjW2O8T2xeI"
Vary: Accept-Encoding
Age: 0
X-Cache: MISS
Accept-Ranges: bytes

{"id":51438805,(snip)...
```

So requests 1, 4 and 5 had empty responses and 2, 3 and 6 work as expected. I've truncated the output here but if you need the full log, I can provide that. I also have the JSON from doing a GET /v1/observations for these 6 observations. I've deleted the observations from iNat because I don't want to pollute the system with these observations.

I actually ran this bash script *as* I was writing it and only saw 1 empty response out of 6 but I don't have the logs for that. So it's important to note that I don't know how to force the issue to happen on any single request, which is no doubt going to be annoying to debug.

To date I haven't seen this issue happen with a "plain" `POST /v1/observations` (i.e. not also passing obs fields) so this whole thing might just be a case of "I told you not to use the undocumented features". And I'll accept that. Just wanted to report it in case it's highlighting something deeper.

I have two dev iNat systems at the moment, the second being a very recent clone of the first, and they don't even behave the same with regards to this issue. They do have different CPU and memory specs however, which makes me feel like maybe this is a race condition issue.

By the way, do you have a development system that I could test out things like this on in future? I'm never sure if problems I see are related to how I've configured my iNat instance and I feel like I should avoid polluting the prod instance where possible.

CC @tokmakoff

**Comment by tomsaleeba** [comment]
I've been working on trying to figure out this issue for a few days and I don't want to run my test harness against real iNat again, so I needed to get that original comment posted that had the results in it.

I have more results and a new test harness though. First, the new harness:

```bash
jwt=${THE_JWT:?you must pass a JWT}
url=https://dev.api.inat.techotom.com/v1/observations

runCount=10
speciesGuess="life"

function randDate {
  echo $(shuf -n1 -i$(date -d '1970-01-02' '+%s')-$(date -d '2020-01-01' '+%s') \
    | xargs -I{} date -d '@{}' '+%Y-%m-%dT00:00:00.000Z')
}

function randLat {
  wholePart=$(shuf -n1 -i11-43)
  decimalPart=$(shuf -n1 -i0-999999)
  echo "-${wholePart}.${decimalPart}"
}

function randLng {
  wholePart=$(shuf -n1 -i112-154)
  decimalPart=$(shuf -n1 -i0-999999)
  echo "-${wholePart}.${decimalPart}"
}

function doCurl {
  lat=$(randLat)
  lng=$(randLng)
  theDate=$(randDate)
  uuid=$(uuidgen)
  echo "
    lat=$lat
    lng=$lng
    date=$theDate
    uuid=$uuid
  "
  # comment either of these lines to exclude from the request
  # projectFragment=',"project_id":4' # wow-dev2
  obsFieldValuesFragment=',"observation_field_values_attributes":{"0":{"observation_field_id":39,"value":"Terrestrial"},"1":{"observation_field_id":51,"value":"Exact"},"2":{"observation_field_id":53,"value":1}}'

  cat <<EOF | curl -X POST --include -H Authorization:$jwt -H 'Content-Type: application/json' $url --data @/dev/stdin
{"ignore_photos":true${projectFragment},"observation":{"latitude":${lat},"longitude":${lng},"observed_on_string":"${theDate}","species_guess":"${speciesGuess}"${obsFieldValuesFragment},"description":null,"positional_accuracy":null,"captive_flag":false,"geoprivacy":"obscured","uuid":"${uuid}"}}
EOF
}

echo "Using URL: $url"

for (( i = 1; i <= $runCount; i++ )); do
  echo -e "\n\n run #$i"
  doCurl $lat $lng $month
done
```

In this configuration, that is not including a project link in the request, it seems to fairly reliably have 2 empty responses out of 10 for this iNat instance. Feel free to sign up to this iNat instance and grab an API key. I can provide SSH access if you require too. Run the harness against my instance as much as you like.

I've found the problem to reproduce itself most reliably with the following conditions (which the above test harness embodies):
  1. include observation field values in the request
  1. do *not* include a project link in the request

In my main iNat instance, including a project linkage seems to make the request always have content. In the other instance, it makes the occurrence of empty responses less common but they are still there.

**Comment by kueda** [comment]
I was able to recreate this locally but haven't quite gotten to the bottom of it yet. It seems like sometimes `Observation.preloadInto` (https://github.com/inaturalist/iNaturalistAPI/blob/master/lib/controllers/v1/observations_controller.js#L30) doesn't actually get the observation out of the index, which sort of suggests the node app is passing the request to the Rails app, which is creating the obs in the db and returning a response to the node app, which then tries to load the new obs out of elasticsearch, but isn't finding it there b/c it hasn't been indexed yet. We are supposed to be guarding against this possibility at https://github.com/inaturalist/inaturalist/blame/master/app/controllers/observations_controller.rb#L652, though, so I'm still confused. Ringing any bells for you, @pleary?

**Comment by pleary** [comment]
I think I see the issue. The things you pointed to @kueda are indeed the issues on the Node API end, but the bug seems to be in rails.

The main problem is we're using Elasticsearch (ES) to store data we fetch for API responses and by default ES only refreshes the data used for querying every 1 second unless told otherwise. What we do is try and wait for the next ES refresh before returning from Observation create requests to make sure the Node API will be able to fetch the observation data from the index. It looks like there are certain places that are overriding the attribute on observations used to indicate we want to wait for a refresh (e.g. https://github.com/inaturalist/inaturalist/blob/master/app/models/observation_field_value.rb#L188).

I can fix the few cases of these to never set `observation.wait_for_refresh` to false if it was already set to true. I also want to update the controller a little to do the observation indexing after all data is created because as it stands it'll index the observation initially, then create things like observation fields. We don't want users to have to wait for a refresh after each observation field is created, which can add seconds to the response time - we should just wait for the refresh once after all data is in place.

**Comment by pleary** [comment]
I think this should be resolved by inaturalist/inaturalist@9f7f0a1 . That should ensure that all observation associations are finished indexing before responding to the observation create request. It has not yet been deployed to production, but I'm seeing large batches working consistently locally

**Comment by tomsaleeba** [comment]
Just deployed the commit to my server and hit it with 100 requests and all worked :tada: 

Thanks @pleary 

I'll close this issue as the fix will hit production at some stage.
